### PR TITLE
feat(PT-5116): optimistic lock handling

### PR DESCRIPTION
**Pull request by dmitry-saraykin**

### CODE REVIEW

File: entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/StreamFullReduceService.kt [hunk 1]
```diff
@@ -25,33 +26,67 @@ open class StreamFullReduceService<Id, Event, E : Identifiable<Id>>(
...
 
     override fun reduce(events: Flow<Event>): Flow<E> = flow {
-        var current: E? = null
-        var entity: E? = null
+        reduceInternal(events = events, initialEntity = null, initialCurrent = null)
+    }
+
+    override fun reduceFromState(initialState: E?, events: Flow<Event>): Flow<E> = flow {
+        val current = initialState?.let {
+            val initialId = entityIdService.getEntityId(initialState)
```
Comment (yehohanan7) [inline entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/StreamFullReduceService.kt:34]: I think E is `Identifiable`, may be we could just get the id directly? `initialState.id`

File: entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/ReduceService.kt [hunk 2]
```diff
@@ -4,4 +4,5 @@ import kotlinx.coroutines.flow.Flow
 
 interface ReduceService<Id, Event, E> {
     fun reduce(events: Flow<Event>): Flow<E>
+    fun reduceFromState(initialState: E?, events: Flow<Event>): Flow<E>
```
Comment (yehohanan7) [inline entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/ReduceService.kt:7]: suggestion: what do you think about just extending the existing `reduce` with an optional initial state? we can bump the major version of the library.

File: entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/StreamFullReduceService.kt [hunk 3]
```diff
@@ -12,7 +13,7 @@ import kotlinx.coroutines.flow.flow
  */
 open class StreamFullReduceService<Id, Event, E : Identifiable<Id>>(
```
Comment (yehohanan7) [inline entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/StreamFullReduceService.kt:14]: I couldn't figure out how the full reduce service will safely do the reduction while the real time reduction is happening at the same time, will the data become inconsistent?

Comment (dmitry-saraykin) [reply entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/StreamFullReduceService.kt:14]: it is done using optimistic locking

File: entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/ReduceService.kt (hunk 2 above)
Comment (dmitry-saraykin) [reply entity-reducer/src/main/kotlin/com/rarible/core/entity/reducer/service/ReduceService.kt:7]: it is not allowed to set default values for interface method parameters so everywhere where we don't need initial state we have to pass null. I think better to have separate methods

**Review by yehohanan7 — APPROVED** [review]
Nice work!

**Pull request merged**